tanjun/_internal/localisation.py

```diff
@@ -0,0 +1,315 @@
+    hikari.CommandType.MESSAGE: "MESSAGE",
+}
+_NamedFields = typing.Literal["check", "option.description", "option.name"]
+_UnnamedFields = typing.Literal["description", "name"]
+_FieldType = typing.Union[_NamedFields, _UnnamedFields]
+
+
+@typing.overload
+def to_localise_id(
+    command_type: typing.Literal["SLASH", "USER", "MESSAGE"],
+    command_name: str,
+    field_type: _NamedFields,
+    field_name: str,
+    /,
+) -> str:
+    ...
+
+
+@typing.overload
+def to_localise_id(
+    command_type: typing.Literal["SLASH", "USER", "MESSAGE"],
+    command_name: str,
+    field_type: _UnnamedFields,
+    field_name: typing.Literal[None] = None,
+    /,
+) -> str:
+    ...
+
+
+def to_localise_id(
+    command_type: typing.Literal["SLASH", "USER", "MESSAGE"],
+    command_name: str,
+    field_type: _FieldType,
+    field_name: typing.Optional[str] = None,
+    /,
+) -> str:
+    """Generate an ID for a localised field.
+
+    Parameters
+    ----------
+    command_type
+        The type of command this field is attached to.
+    command_name
+        Name of the command this field is attached to.
+    field_type
+        The type of field this localisation ID is for.
+    field_name
+        Name of the field this localisation ID is for.
+
+        This doesn't apply to command names and descriptions.
+
+    Returns
+    -------
+    str
+        The generated localied field ID.
+    """
+    if field_name:
+        if field_type == "name" or field_type == "description":
+            raise RuntimeError(f"Field_name must not be provided for {field_type} fields")
+
+        return f"{command_type}:{command_name}:{field_type}:{field_name}"
+
+    if field_type != "name" and field_type != "description":
+        raise RuntimeError(f"Field_name must be provided for {field_type} fields")
+
+    return f"{command_type}:{command_name}:{field_type}"
+
+
+def localise_command(cmd_builder: hikari.api.CommandBuilder, localiser: dependencies.AbstractLocaliser, /) -> None:
+    """Localise the fields for a command builder.
+
+    Parameters
+    ----------
+    cmd_builder
+        The application command builder to localise the fields for.
+    localiser
+        The abstract localiser to localise fields with.
+    """
+    localise_type = _TYPE_TO_STR[cmd_builder.type]
+    names = dict(cmd_builder.name_localizations)
+    names.update(localiser.get_all_variants(to_localise_id(localise_type, cmd_builder.name, "name")))
+    cmd_builder.set_name_localizations(names)
+
+    if isinstance(cmd_builder, hikari.api.SlashCommandBuilder):
+        descriptions = dict(cmd_builder.description_localizations)
+        descriptions.update(localiser.get_all_variants(to_localise_id(localise_type, cmd_builder.name, "description")))
+        cmd_builder.set_description_localizations(descriptions)
+
+        name = cmd_builder.name
+        options = cmd_builder.options
+        while options:
+            if options[0].type in _internal.COMMAND_OPTION_TYPES:
```

Comment:
This is wrong, you styled this like command search execution but a command group can have multiple commands in jt